controllers/flinkcluster_updater.go

```diff
@@ -594,22 +618,58 @@ func (updater *ClusterStatusUpdater) deriveClusterStatus(
 				}
 				// Savepoint for job-cancel
 				var observedSavepoint = observed.cluster.Status.Savepoint
-				if observedSavepoint == nil || observedSavepoint.State != v1beta1.SavepointStateInProgress {
+				if observedSavepoint == nil ||
+					(observedSavepoint.State != v1beta1.SavepointStateInProgress && observedSavepoint.State != v1beta1.SavepointStateNotTriggered) {
 					updater.log.Info("There is no savepoint in progress. Trigger savepoint in reconciler.")
-					status.Savepoint = &v1beta1.SavepointStatus{State: v1beta1.SavepointStateNotTriggered, TriggerReason: v1beta1.SavepointTriggerReasonJobCancel}
+					status.Savepoint = &v1beta1.SavepointStatus{
+						State:         v1beta1.SavepointStateNotTriggered,
+						TriggerReason: v1beta1.SavepointTriggerReasonJobCancel,
+					}
 				} else {
 					updater.log.Info("There is a savepoint in progress. Skip new savepoint.")
 				}
 				controlStatus = getNewUserControlStatus(userControl)
 			}
 		}
 	}
-	// maintain control status if there is no change
+	// Maintain control status if there is no change.
 	if recorded.Control != nil && controlStatus == nil {
 		controlStatus = recorded.Control.DeepCopy()
 	}
 	status.Control = controlStatus
 
+	// Handle update.
+	var savepointForJobUpdate *v1beta1.SavepointStatus
+	if isUpdateTriggered(*recorded) {
+		switch getUpdateState(*observed) {
+		case UpdateStateStoppingJob:
+			// Even if savepoint has been created for update already, we check the age of savepoint continually.
+			// If created savepoint is old and savepoint can be triggered, we should take savepoint again.
+			// (e.g., for the case update is not progressed by accidents like network partition)
+			if !isSavepointUpToDate(observed.observeTime, *jobStatus) &&
+				canTakeSavepoint(*observed.cluster) &&
+				(recorded.Savepoint == nil || recorded.Savepoint.State != v1beta1.SavepointStateNotTriggered) {
+				savepointForJobUpdate = &v1beta1.SavepointStatus{
+					State:         v1beta1.SavepointStateNotTriggered,
+					TriggerReason: v1beta1.SavepointTriggerReasonJobUpdate,
+				}
+				updater.log.Info("Savepoint will be triggered for job update")
+			} else if recorded.Savepoint != nil && recorded.Savepoint.State == v1beta1.SavepointStateInProgress {
+				updater.log.Info("Savepoint is in progress")
+			} else {
+				updater.log.Info("Stopping job")
+			}
+		case UpdateStateUpdating:
```

Comment:
Does this state include both components update and job update? How do you ensure the order of them?

Reply:
Fixed also via `isClusterReady` check in reconciler.